_posts/2020-08-24-native-loader-net5.md

```diff
@@ -0,0 +1,38 @@
+---
+layout: blog
+title: "Native Library Loading in .NET 5"
+author: Ryan Lucia
+tags: [loader]
+---
+
+After years of work, Mono can now be built out of the [dotnet/runtime repository](https://github.com/dotnet/runtime) in a .NET 5-compatible mode! This mode means numerous changes in the available APIs, managed and embedding, as well as internal runtime behavioral changes to better align Mono with CoreCLR and the .NET ecosystem.
+
+One area with multiple highly impactful changes to the runtime internals is library loading. For managed assemblies, Mono now follows the algorithms outlined on [this page](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-managed), which result from the removal of `AppDomain`s and the new `AssemblyLoadContext` APIs. The only exception to this is that Mono still supports bundles registered via the embedding API, and so the runtime will check that after the active `AssemblyLoadContext` cache.
+
+The managed loading changes are fairly clear and well documented, but unmanaged library loading has changed in numerous ways, some of them far more subtle.
+
+First off, Mono now follows [this algorithm](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-unmanaged) for P/Invoke resolution. Notably absent here is DllMap's XML config files, which have been disabled entirely in .NET 5. The DllMap embedding APIs are also disabled by default on desktop platforms, but can be enabled by building with `--enable-minimal=dllmap`.
```

Comment:
@CoffeeFlux This doesn't look right. I don't think `--enable-minimal=dllmap` will *bring back* dllmap support.

`--enable-minimal=foo` usually _disables_ support for feature *foo* by defining `DISABLE_FOO`.   And as far as I know there's no `configure` thing to turn on a disabled minimal feature that's off by default.

Looking at `configure.ac` I think the only way to turn dllmap back on is to hack the .ac file or not to build with the `netcore` preset.

I think it might be better to just leave it at "We dropped support for dllmap in net5".

Reply:
Oh yeah, oops. Works for me.